### PR TITLE
fix(tokenless): skip TOON encoding for payloads under 500 chars

### PR DESCRIPTION
**Pull request by Forrest-ly**
## 改动说明

本 PR 包含两个独立修复（分属两个 commit）。

### 1. TOON 编码最小负载阈值（500 字符）

TOON 编码此前对每个 PostToolUse 负载都会执行。对典型的小型工具输出（<500 字符的 JSON），TOON 平均只能节省个位数字符（约 0.3%），却每次都要付出完整的编码计算开销，并在统计中产生大量近零收益记录，压缩率反馈被稀释。

本 PR 为全部支持 TOON 的管线增加最小负载阈值（500 字符，按 Unicode 字符数 / code point 计数）：

- `compress_response_hook.py`（响应压缩 + TOON 组合管线）：压缩结果 ≥500 字符才进入 TOON 步骤；更小的负载保留响应压缩结果，完全不调用 `compress-toon`（不产生子进程、不编码、不写统计）。
- `compress_toon_hook.py`（独立 TOON hook）：最小响应阈值由 200 提升到 500 字符，原因相同。
- `hermes/__init__.py`、`openclaw/index.ts`、`codex/scripts/compress-response`（adapter 专属管线）：同样应用 500 字符守卫，门槛以下不拉起 `compress-toon` 子进程。
- 计数语义统一：Python hooks 对结构化负载以 `ensure_ascii=False` 规范化后计长（`\u` 转义不膨胀计数）；OpenClaw 按 Unicode code point 计数（`hasAtLeastChars`，代理对计 1 次）；与 JSON 字符串输入路径一致。
- 阈值以下不发生 `compress-toon` 调用，因此不再产生近零收益的统计记录——统计中保留的即为有效压缩，无需在数据库中额外区分跳过情形（保持 DB/diff/SLS 语义不变）。
- 用户文档（en/zh framework-integration）补充阈值说明（注明 500 为当前实现常量、后续可能调整，且覆盖全部 TOON 管线）。
- 新增共享 hooks 与 Hermes、Codex、OpenClaw adapter 的回归测试（call-marker mock 证明阈值以下不调用 compress-toon、阈值以上仍调用；含结构化非 BMP 边界用例），并注册进 `make test-integration` / `make test-adapters`。

直接调用 `tokenless compress-toon` CLI 的语义保持不变（显式调用属用户主动行为，且已有"无收益则输出原文"的保护）。

### 2. npm 平台包 bin 链接修复（独立 commit + 版本号 bump）

`Fixes: 5068b6fe ("feat(tokenless): add npm package support for anolisa-tokenless")`
Related to AGE-3475（CI-Fix #2613）

**根因**：平台包（`@anolisa/tokenless-*`）与根包声明了同名 bin 入口（`tokenless`/`rtk`/`toon`）。当 tree 中多个包声明同名 bin 时，npm（至少到 10.x）的 reify 会删除所有冲突的 `.bin` 链接而不是择一保留。0.7.7 发布到 registry 之后，CI 冒烟测试安装本地 tarball、以及终端用户 `npm install anolisa-tokenless@0.7.7` 都会触发：根包的 optionalDependencies packument（已发布、声明同名 bin）进入 ideal tree，安装结果 `.bin/` 中没有 `tokenless` 可执行文件（exit 127）。这既导致本 PR 的 CI 失败，也使已发布的 0.7.7 对终端用户不可用。

**修复**（与 esbuild 的 `@esbuild/<platform>` 包设计一致）：
- 平台包不再声明 bin（二进制仍放在 `bin/` 目录）；根包独占三个 bin 入口，其 postinstall 将它们链接到已安装平台包内的原生二进制。
- `tests/test-package-npm-prebuilt.sh` 新增断言：平台 manifest 不得包含 bin 条目，防止回归。
- 版本号 0.7.7 → 0.7.8（独立 chore commit，同步 Cargo.toml / component.toml / Cargo.lock×3 / npm manifests / spec.in / CHANGELOG）：0.7.7 已发布且 npm 禁止重复发布，其 registry packument 仍声明 bin；bump 到 0.7.8 使 CI 冒烟测试摆脱已发布 0.7.7 的 registry 状态干扰，下一次发布即可让终端用户拿到修复。
- CI 冒烟测试**不**引入 `npm rebuild` 之类的绕过：冒烟测试的价值在于忠实复现终端用户安装路径，rebuild 会掩盖此类打包损坏；回归由上述打包断言守护。

**验证**：本地以真实源码运行 `make test-npm-package`（全部断言通过），并按 CI 冒烟序列在干净目录安装本地构建的 platform + root tarball：`.bin/tokenless → anolisa-tokenless/bin/tokenless → @anolisa/tokenless-linux-x64/bin/tokenless` 链接链完整，postinstall 正确建立符号链接，openclaw adapter 产物齐全。推送后 CI run 32105553498 全绿（Test tokenless / Check component versions / Test anolisa 通过，0 失败）。

> 分支历史说明：分支已 rebase 到最新 main（包含 main 上的 0.7.7 版本 bump），共 3 个 commit（原始修复 + bin 修复 + 版本 bump）。此前的并行修复 commit（f90fb0ca / 50076e43，方案为 bin 移除 + ci.yaml `npm rebuild` 绕过）已被当前版本取代：当前版本不依赖 registry 状态、不削弱冒烟测试，且为下一次发布准备好了版本号。

## 风险和兼容性

- TOON 阈值：仅影响 <500 字符负载（跳过近零收益的 TOON 步骤），≥500 字符负载行为不变；统计记录减少属预期效果。回滚：还原对应 commit 即可，无数据迁移。
- npm 平台包：bin 字段移除后平台包不再提供可直接执行的 bin 条目，入口统一为根包 postinstall 链接；未发布平台包的环境不受影响。回滚：还原该 commit（注意会重新引入 bin 冲突问题）。
- 版本 bump（0.7.8）：纯版本号同步（Cargo/RPM/npm manifest/CHANGELOG），无代码逻辑变更；如需回滚连同 bin 修复一起还原即可。0.7.7 已发布在 registry 且无法覆盖，勿再基于 0.7.7 发布。

## 测试情况

**环境**：Linux x86_64；rustc/cargo 1.94.1；Python 3.11；Node v22.21.1；tokenless 0.7.6（本地源码 release 构建）。

**实际执行的命令与结果**

| 项目 | 命令 | 结果 |
| --- | --- | --- |
| 格式 | `cargo fmt --all -- --check` | 通过 |
| Lint | `cargo clippy --workspace -- -D warnings` | 0 错误 |
| Rust 测试 | `cargo test --workspace -- --test-threads=1` | 573 通过 / 0 失败 / 2 忽略 |
| Python 集成 | `test_compress_response_hook`（21 用例，含新增 3 个结构化非 BMP 阈值用例）、`test_compress_toon_hook`（4 用例，含新增 2 个结构化非 BMP 用例）、`test_compress_schema_hook`、`test_agentscope_middleware`、`test_agentscope_v1`、`test_resolve_agent_id` | 全部 OK |
| 阈值回归 | `test_codex_toon_threshold`、`test_hermes_toon_threshold`、`test-openclaw-toon-threshold.mjs`（4 用例，含 2 个非 BMP 边界用例） | 全部通过 |
| 功能 hook 套件 | `make test-hooks`（含 run-all-tests.sh，使用本次源码构建的 release 二进制） | 72/72 通过 |
| OpenClaw 插件 | `make build-openclaw-plugin`（tsc）+ `node --test tests/test-openclaw-rtk-context.mjs` | 构建通过 / 6/6 通过 |
| npm 打包 | `make test-npm-package`（含平台 manifest 无 bin 断言） | 通过 |

**修复验证（复现 → 验证）**
- Unicode 计数（复现 → 修复）：结构化 `{"stdout": "😀"×40}`（67 个 Unicode 字符，`\u` 转义序列化后 507 字符）修复前两个 Python hook 均会拉起 compress-toon，修复后直接跳过（call-marker 证明，负向对照成立）；OpenClaw 的 244-emoji 复现样例同样不再拉起。
- 真实二进制端到端：869 字符 JSON 经独立 TOON hook 输出 TOON 表格格式；293 字符 JSON 被直接跳过（空输出、无编码调用）。组合管线：1444 字符负载产出 TOON 表格输出；335 字符负载仅做响应压缩、跳过 TOON 步骤。
- npm bin 链接：平台 manifest 不再声明 bin；`make test-npm-package` 全部断言通过。

**未运行项及原因**
- `tests/test_rewrite_hook.py` 与 `tests/test_hermes_plugin_import.py::test_resolves_via_xdg_data_home`：测试机存在系统级安装的 tokenless/rtk（/usr/local 前缀），干扰测试沙箱；未改动基线上同样复现，与本次改动无关。
- `tests/test-toon.sh` / `test-toon-full.sh`：需要完整安装环境（toon/openclaw/已安装 hook 路径），本环境不具备全部依赖；本次未改变 CLI 语义，不受影响（未运行）。

## 文档和回滚

- 文档更新：`docs/user-guide/{en,zh}/token-saving/tokenless/framework-integration.md`（TOON 阈值说明）。
- 回滚：三个 commit（TOON 阈值 / npm bin 修复 / 版本 bump）相互独立，可分别 revert；TOON 阈值无数据迁移，npm 修复 revert 会重新引入 bin 冲突问题。

### CODE REVIEW

File: src/tokenless/adapters/tokenless/common/hooks/compress_toon_hook.py [hunk 1]
```diff
@@ -39,7 +39,11 @@
 # -- constants ---------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (Forrest-ly) [reply src/tokenless/adapters/tokenless/common/hooks/compress_toon_hook.py:39]: 已处理（5f36690b）：

- 文档：en/zh framework-integration 现在明确说明 500 是"当前实现的阈值，后续可能调整"（the current implementation threshold, which may be adjusted later），并注明该阈值适用于所有支持 TOON 的管线。
- 测试：`test_compress_toon_hook.py` 模块 docstring 与 `test_compress_response_hook.py` 阈值测试类 docstring 均已引用 `_MIN_TOON_CHARS`；本次新增的 Hermes/Codex/OpenClaw 阈值回归测试同样直接引用常量名（`_MIN_TOON_CHARS` / `MIN_TOON_CHARS`）。

File: docs/user-guide/en/token-saving/tokenless/framework-integration.md [hunk 2]
```diff
@@ -39,7 +39,7 @@ The standalone `compress-response` defaults are not the defaults used by most ad
 | Shell/exec | 65,536-character strings, 128 retained array items, depth 8 |
 | Other structured tools | 1,048,576-character strings, 65,536 retained array items, depth 32 |
 
-The shared response hook, OpenClaw, and Hermes skip inputs shorter than 200 characters. Codex skips inputs shorter than 500 characters; it includes compressed content only for inputs of at least 4,000 characters and otherwise adds diagnostics or a summary. Skill-like text with YAML frontmatter is also skipped by the shared paths.
+The shared response hook, OpenClaw, and Hermes skip inputs shorter than 200 characters. Codex skips inputs shorter than 500 characters; it includes compressed content only for inputs of at least 4,000 characters and otherwise adds diagnostics or a summary. Skill-like text with YAML frontmatter is also skipped by the shared paths. The TOON encoding step only runs on payloads of at least 500 characters; smaller payloads keep the compressed form, because TOON savings on small JSON are negligible.
```
[the remark this answers is not in the document]
Comment (Forrest-ly) [reply docs/user-guide/en/token-saving/tokenless/framework-integration.md]: Fixed in 5f36690b — applied the 500-char guard to every TOON-capable adapter pipeline instead of narrowing the docs:

- `adapters/tokenless/hermes/__init__.py`: the TOON step is skipped when the current (possibly response-compressed) payload is under `_MIN_TOON_CHARS` (500); such payloads keep the response-compressed form. The 200-char `_MIN_RESPONSE_LEN` entry gate is unchanged.
- `adapters/tokenless/openclaw/index.ts`: `tryCompressToon` returns null before spawning `compress-toon` when `JSON.stringify(response)` is under `MIN_TOON_CHARS` (500). The 200-char entry gate is unchanged.
- `adapters/tokenless/codex/scripts/compress-response`: same guard on the TOON input, covering the case where response compression shrinks a >=500-char input below the threshold.

So no adapter spawns `compress-toon` (or records near-zero stats) for sub-500-char payloads anymore, and the documented statement now holds for the shared hooks, Codex, OpenClaw, and Hermes alike.

Regression tests added for all three adapters (call-marker mocks prove `compress-toon` is not invoked below the threshold and still runs at/above it): `tests/test_hermes_toon_threshold.py`, `tests/test_codex_toon_threshold.py` (registered in `make test-integration`), and `tests/test-openclaw-toon-threshold.mjs` (registered in `make test-adapters`; negative control: the small-payload test fails without the guard).

File: src/tokenless/adapters/tokenless/openclaw/index.ts [hunk 3]
```diff
@@ -326,6 +332,9 @@ function tryCompressResponse(response: any, sessionId?: string, toolCallId?: str
 function tryCompressToon(response: any, sessionId?: string, toolCallId?: string): { toonText: string; savingsPct: number } | null {
   try {
     const input = JSON.stringify(response);
+    // Skip payloads below the minimum threshold: TOON savings on small
+    // JSON are near-zero but the encode cost is paid on every tool result.
+    if (input.length < MIN_TOON_CHARS) return null;
```
Comment (ikunkun-sys) [inline src/tokenless/adapters/tokenless/openclaw/index.ts]: **[P2] Count payload characters consistently across adapters**

This guard uses JavaScript `String.length` (UTF-16 code units), although the documented threshold is in characters and the Python paths use Unicode code-point counts. At this head, `JSON.stringify({stdout: "😀".repeat(244)})` contains 257 Unicode code points but has `input.length === 501`; an instrumented run showed `spawnSync` invoking `compress-toon` and changing the message. Thus an under-500-character non-BMP payload still pays the child-process cost and can generate the near-zero-savings records this change is intended to remove. Define one counting unit for every adapter and cover this boundary with a non-BMP regression case.

**Review by ikunkun-sys — CHANGES_REQUESTED** [review]
Requesting changes for two issues verified at current head `5f36690b1c6b`:

1. The OpenClaw TOON gate counts UTF-16 code units, so some payloads with fewer than 500 Unicode characters still spawn `compress-toon`; see the inline finding.
2. Commit `5f36690b` is a review follow-up that fixes adapter coverage introduced by `a7b6cff7` in this PR. Per AGENTS.md §13, it should be fixup/amended into the originating commit and autosquashed before merge.

Validation: `make test-integration`, the OpenClaw TypeScript build, the new OpenClaw threshold tests, RTK regression tests, and `git diff --check` passed. A local synthetic merge of current `main` (`d80cd127`) with this head was conflict-free and passed the same targeted gates.

File: src/tokenless/adapters/tokenless/openclaw/index.ts (hunk 3 above)
Comment (Forrest-ly) [reply src/tokenless/adapters/tokenless/openclaw/index.ts]: Both points addressed — the branch is now a single commit `f580cdda` (the earlier fixup commit was squashed into the originating commit per AGENTS.md §13, and the force-push kept the branch mergeable).

**1. Counting unit — adopted.** `tryCompressToon` now gates on Unicode code points instead of UTF-16 code units: a new `hasAtLeastChars()` helper iterates the string (a surrogate pair counts once, matching the Python adapters' `len()` semantics) and returns as soon as the threshold is reached, so large payloads only pay for scanning the first 500 characters. Your repro case (244 × 😀 ≈ 271 code points but >500 UTF-16 units) no longer spawns `compress-toon`. Two non-BMP boundary cases were added to `tests/test-openclaw-toon-threshold.mjs`: the under-threshold emoji payload must skip `compress-toon`, and an at/above-threshold emoji payload (~547 code points) must still encode. The other adapters already count characters (Python `len()` = code points), so all TOON-capable pipelines now share one counting unit for the threshold. The pre-existing 200-char entry gate for response compression keeps its prior semantics as it predates and is out of scope for this change.

**2. Commit hygiene — adopted.** `5f36690b` was a follow-up fixing coverage introduced by `a7b6cff7` in this same PR, so it has been amended into the originating commit; no standalone fix commit remains.

Validation at `f580cdda`: OpenClaw plugin rebuilt with `tsc`, `node --test tests/test-openclaw-toon-threshold.mjs` 4/4 pass (including both non-BMP cases), `tests/test-openclaw-rtk-context.mjs` 6/6 pass, the Python threshold/hook suites (`test_compress_response_hook`, `test_compress_toon_hook`, `test_codex_toon_threshold`, `test_hermes_toon_threshold`) all pass, and `git diff --check` is clean.

**Review by kongche-jbw — COMMENTED** [review]
Review baseline: `d80cd127c0f6...f90fb0cac3aa`

**[P2] 统一 Python hook 对结构化 Unicode 负载的字符计数**

`compress_response_hook.py:322,394` 与 `compress_toon_hook.py:91,100` 会先用默认
`json.dumps` 把非 ASCII 转成 `\u` 转义，再对转义后的字符串执行 500 字符门限。
例如结构化 `{"stdout": "😀" × 40, ...}` 解码后只有 67 个字符，但序列化后为 507
个字符；两个生产 hook 都会调用 `compress-toon`，与同内容的 JSON 字符串输入及
OpenClaw 的 Unicode code-point 语义不一致，并重新产生本改动要消除的子进程和统计。
Possible direction: 对门限计数使用不转义的规范化 JSON，或直接按解码值计算，并为
结构化非 BMP 输入补充两个 hook 的回归测试。

**[P2] 补齐追加 npm 修复的提交归属与 PR 说明**

`.github/workflows/ci.yaml:910` 及平台包 manifest 属于第二个独立 npm 修复，但当前
PR 标题、说明、风险和验证仍只描述 TOON 门限。该提交修复的是 `5068b6fe` 已在
`main` 引入的问题，也未按 `AGENTS.md` §13 写入 `Fixes:` 归属。这会让审查和后续
回滚看不到实际包装行为变化。
Possible direction: 在提交正文补 `Fixes: 5068b6fe (...)`，并更新 PR 说明，或拆分 PR。

File: src/tokenless/adapters/tokenless/common/hooks/compress_toon_hook.py [hunk 4]
```diff
@@ -91,8 +95,9 @@ def main() -> None:
     if not tool_response:
         skip()
 
-    # 7. Skip small responses (character count, not byte length)
-    if len(tool_response) < _MIN_RESPONSE_CHARS:
+    # 7. Skip payloads below the TOON minimum threshold (character count,
+    # not byte length): TOON savings on small JSON are near-zero
+    if len(tool_response) < _MIN_TOON_CHARS:
```
Comment (ikunkun-sys) [inline src/tokenless/adapters/tokenless/common/hooks/compress_toon_hook.py:106]: **[P1] Gate structured payloads by Unicode characters**

The dict/list branch above serializes with Python's default `ensure_ascii=True`, so this length gate counts `\uXXXX` escape text rather than decoded Unicode code points. At this head, `{"stdout": "😀" × 40, ...}` is about 67 code points but 507 serialized characters; a marker probe shows that this hook and the sibling `compress_response_hook.py` path both invoke `compress-toon`. This violates the documented sub-500-character skip contract and restores the child-process/statistics noise the PR is intended to remove. Preserve the code-point unit for structured payloads and cover both Python hooks with non-BMP regression cases.

**Review by ikunkun-sys — CHANGES_REQUESTED** [review]
Requesting changes at current head `4594701576f9` after the latest force-push. The earlier OpenClaw code-point and autosquash findings are resolved, but two current-head blockers remain:

1. Structured Python payloads still use the escaped `json.dumps` representation for the size gate; see the inline finding. A marker probe confirmed that both the standalone and combined hooks invoke `compress-toon` for `{"stdout": "😀" × 40, ...}` even though the decoded payload is only about 67 Unicode code points.
2. Commit `ccd97b03` fixes the npm bin collision introduced by main commit `5068b6fe`, but it has no `Fixes:` trailer (`git interpret-trailers --parse` is empty), contrary to AGENTS.md §13. The PR description also says this trailer and `ensure_ascii=False`/structured-Unicode tests are present, and still reports the old 0.7.6 and 21/4 test evidence, while this head is 0.7.8 with 18/2 Python hook cases. Please synchronize the commit attribution and PR evidence with the actual head.

Validation at this SHA: the Unicode marker probe reproduces both unwanted calls; targeted integration, OpenClaw, and npm packaging tests pass; `git diff --check` is clean; and current GitHub CI is green. The PR is MERGEABLE but remains BLOCKED/CHANGES_REQUESTED.

File: src/tokenless/adapters/tokenless/common/hooks/compress_toon_hook.py (hunk 4 above)
Comment (Forrest-ly) [reply src/tokenless/adapters/tokenless/common/hooks/compress_toon_hook.py:106]: 问题属实——已修复，head 更新为 `136e5c1e`（保持 MERGEABLE）。

**原因**：上一轮修复（`50076e43`）之后的一次分支重写（rebase 到新基线并追加版本号提交）把两个 Python hook 的修复从 TOON 提交中丢掉了；OpenClaw 的 code point 修复幸存，因此只暴露出 Python 侧问题。

**本次处理**：

- 恢复 `compress_toon_hook.py` 与 `compress_response_hook.py` 的结构化负载修复：dict/list 分支以 `ensure_ascii=False` 规范化后再计长，门限按 Unicode code point 统计，与 JSON 字符串输入路径及 OpenClaw（`hasAtLeastChars`）一致；`_build_replacement_output` 的收益比较同步统一。按 AGENTS.md §13 fixup 进 TOON 原提交，不保留独立 fix commit。
- 恢复两个 hook 的结构化非 BMP 回归用例：`{"stdout": "😀"×40}`（约 67 code points、转义后 507 字符，即本条意见的复现样例）必须跳过 compress-toon；200-emoji 用例（227 字符，通过 200 字符入口门限以隔离 TOON 门限）同样跳过；520 emoji（547 字符）照常编码。负向对照：跳过类用例在修复前代码上失败。
- 同一次重写还丢掉了 npm 修复提交的 `Fixes: 5068b6fe ("feat(tokenless): add npm package support for anolisa-tokenless")` 归属（上一轮 review 已确认的要求），一并恢复。

**验证**（head `136e5c1e`）：`test_compress_response_hook` 21 用例全部 OK（含 3 个结构化非 BMP 用例）、`test_compress_toon_hook` 4 用例全部 OK（含 2 个结构化非 BMP 用例）、`test_codex_toon_threshold` / `test_hermes_toon_threshold` OK、`test-openclaw-toon-threshold.mjs` 4/4 通过、`make test-npm-package` 通过、`git diff --check` 干净。

File: src/tokenless/adapters/tokenless/common/hooks/compress_toon_hook.py [hunk 5]
```diff
@@ -84,15 +88,22 @@ def main() -> None:
...
         skip()
 
     if not tool_response:
         skip()
 
-    # 7. Skip small responses (character count, not byte length)
-    if len(tool_response) < _MIN_RESPONSE_CHARS:
+    # 7. Skip payloads below the TOON minimum threshold (character count,
+    # not byte length): TOON savings on small JSON are near-zero
+    if len(tool_response) < _MIN_TOON_CHARS:
```
Comment (ikunkun-sys) [inline src/tokenless/adapters/tokenless/common/hooks/compress_toon_hook.py:106]: **[P1] Keep string-wrapped JSON under the code-point gate**

This gate still receives an ASCII-escaped string when `tool_response_raw` is itself a JSON-encoded string: `unwrap_string_json()` parses the inner object and returns `json.dumps(inner_obj, ...)` with the default `ensure_ascii=True`. At this head, an inner `{"stdout": "😀" × 40, ...}` is 67 code points (the wrapped input is 75), but normalization expands it to 507 characters; marker probes show that both this hook and `compress_response_hook.py` invoke `compress-toon`. Thus the explicitly supported string-wrapped path still violates the documented sub-500-character skip contract. Cover that path in both Python hook suites and keep its normalization in the same code-point unit as dict/list and OpenClaw inputs.

**Review by ikunkun-sys — CHANGES_REQUESTED** [review]
Requesting changes at current head `136e5c1eae24`. The previous dict/list Unicode normalization and npm `Fixes:` attribution blockers are resolved, but one supported input path still violates the same threshold contract:

- A `tool_response` value that is itself a JSON-encoded string goes through `unwrap_string_json()`, which reserializes the decoded object with default `ensure_ascii=True`. The inline finding shows that both Python hooks therefore still invoke `compress-toon` for an under-500-code-point non-BMP payload.

Current targeted integration, OpenClaw, npm packaging, npm 10 link, and `git diff --check` validations pass, and GitHub CI is green. A dedicated marker probe for the wrapped input fails in both Python hooks, confirming that the green suites do not cover this branch.

File: src/tokenless/adapters/tokenless/common/hooks/compress_toon_hook.py (hunk 5 above)
Comment (Forrest-ly) [reply src/tokenless/adapters/tokenless/common/hooks/compress_toon_hook.py:106]: 问题属实——已修复，head 更新为 `31477435`（保持 MERGEABLE）。

**根因**：遗漏分支在共享辅助函数 `unwrap_string_json`（`common/hooks/hook_utils.py`）：string-wrapped 输入解出内层 dict/list 后用默认 `ensure_ascii=True` 重新序列化，gate 收到的仍是转义文本。该 helper 仅被这两个 hook 使用。

**修复**：

- `unwrap_string_json` 的重新序列化改为 `ensure_ascii=False`，string-wrapped 路径与 dict/list 分支及 OpenClaw 输入统一按 Unicode code point 计长。按 AGENTS.md §13 fixup 进 TOON 原提交（`be0e9ea7`），不保留独立 fix commit。
- 两个 hook 测试套件新增 string-wrapped 非 BMP 回归用例：包裹 `{"stdout": "😀"×40}`（内层约 67 code points、包裹输入约 73，即本条意见的复现样例）必须跳过 compress-toon；response hook 另加包裹 200-emoji 用例（解包后 227 字符，通过 200 字符入口门限以隔离 TOON 门限）；包裹 520-emoji（解包后 547 字符）照常编码。负向对照：3 个跳过用例在修复前代码上失败（实测复现了本条意见）。

**最终 head 确认**（应要求复核）：`31477435` 中 TOON 提交 `be0e9ea7` 同时包含 dict/list 分支与 `unwrap_string_json` 两处 `ensure_ascii=False` 修复，两个 hook 测试套件各含结构化与 string-wrapped 两类非 BMP 回归用例（修复后 24 + 6 用例全部通过）；npm 修复提交的 `Fixes: 5068b6fe (...)` 归属仍在。

**验证**（head `31477435`）：`test_compress_response_hook` 24 用例全部 OK、`test_compress_toon_hook` 6 用例全部 OK、`test_codex_toon_threshold` / `test_hermes_toon_threshold` / `test_compress_schema_hook` / `test_agentscope_*` / `test_resolve_agent_id` OK、`test-openclaw-toon-threshold.mjs` 4/4 通过、`git diff --check` 干净。

**Review by ikunkun-sys — APPROVED** [review]
Reviewed current head `314774350551`. The previously requested changes are resolved: OpenClaw and both Python hook paths now apply the TOON threshold using Unicode code points, including structured and string-wrapped JSON inputs, and the npm fix has the required `Fixes:` attribution. Targeted integration, npm packaging, and OpenClaw regression tests passed locally, and the current GitHub checks are green.

Non-blocking: the PR description still lists the earlier Python test counts (21/4 instead of 24/6) and local version 0.7.6 instead of the current 0.7.8 source.
